run-cgprna/run_cgprna/map.py

```diff
@@ -0,0 +1,135 @@
+            for key,value in ref_related_args.items():
+                # GTF file name will be the same as in the bundle
+                if key!= '--gene-build-gtf-name':
+                    if value is None:
+                        ref_related_args[key] = ref_related_defaults[key]
+                else:
+                    if value is not None:
+                        print('Warning: provided "--gene-build-gtf-name" will be overwritten by the GTF file name in the reference bundle.')
+
+            # make the folder structure
+            bundle_decompress_path = os.path.join(temp_dir, ref_root_dir_name, ref_related_args['--species'], ref_related_args['--reference-build'], 'star')
+            final_gtf_folder = os.path.join(bundle_decompress_path, ref_related_args['--gene-build'])
+            mkdir(final_gtf_folder)
+
+            # dump reference bundle
+            untar(args.ref, bundle_decompress_path)
+
+            # find the GTF file
+            gtfs = find('*.gtf', bundle_decompress_path)
+            if len(gtfs) == 1:
+                ref_related_args['--gene-build-gtf-name'] = os.path.basename(gtfs[0])
+                # link the file to final_gtf_folder
+                os.symlink(
+                    gtfs[0],
+                    os.path.join(final_gtf_folder, ref_related_args['--gene-build-gtf-name'])
+                )
+            else:
+                sys.exit('Error: none or too many GTF files in refence bundle. Found GTF(s): %s' % ','.join(gtfs))
+        else:
+            sys.exit('Error: wrong input format. "--reference" can only be a tar.gz file or a folder.')
+    else:
+        if any([ele is None for ele in ref_related_args.values()]):
+            sys.exit(
+                'Error: missing required input. When "--reference" is not a reference bundle tar file, you have to provide: %s' % ', '.join(
+                    [ key for key, value in ref_related_args.items() if value is None])
+            )
+   
+    # gathering parameters
+    params = {
+        **vars(args),
+        'raw_reads_string': ' '.join([os.path.abspath(path) for path in args.input]),
+        'reference_data_root': reference_data_root,
+        'other_options': ' '.join(other_options),
+        'out_dir': os.path.abspath(args.out_dir),  # overwrite the value in args with absolute path
+        'species': ref_related_args['--species'],  # overwrite the value in args
+        'ref_build': ref_related_args['--reference-build'],  # overwrite the value in args
+        'gene_build': ref_related_args['--gene-build'],  # overwrite the value in args
+        'gene_build_gtf_name': ref_related_args['--gene-build-gtf-name']  # overwrite the value in args
+    }
+
+    run_templates_in_shell(
+        [
+            STAR_MAP_TEMPLATE,
+            MARK_DUPS_TEMPLATE,
+            BAM_INDEX_TEMPLATE
+        ],
+        params)
+    
+    # clean temp dir
+    if clean_temp:
+        shutil.rmtree(temp_dir)
+
+
+def find(pattern, path):
```

Comment:
This could be rewritten as a chaining of generators.

Reply:
Tried and love it!